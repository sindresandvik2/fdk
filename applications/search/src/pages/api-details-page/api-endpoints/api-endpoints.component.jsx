import React from 'react';
import PropTypes from 'prop-types';
import _ from 'lodash';
import localization from '../../../lib/localization';

const allowedEndpointOperations = [
  'get',
  'put',
  'post',
  'delete',
  'options',
  'head',
  'patch',
  'trace'
];

export const renderPathMethod = (path, method, methodDeclaration) => (
  <React.Fragment key={`${method}-${path}`}>
    <div className="d-flex list-regular--item">
      <div className="col-5 pl-0 fdk-text-strong">
        {method.toUpperCase()} {path}
      </div>
      <div className="col-7">{methodDeclaration.description}</div>
    </div>
  </React.Fragment>
);

export const renderPath = (path, declaration) => {
  if (!declaration) {
    return null;
  }
  // only display endpoint operations as predefined in constant
  const allowedDeclaration = _.pick(declaration, allowedEndpointOperations);

  return Object.keys(allowedDeclaration).map(method =>
    renderPathMethod(path, method, allowedDeclaration[method])
  );
};

export const renderPaths = paths => {
  const renderedMethods =
    paths && Object.keys(paths).map(path => renderPath(path, paths[path]));
  return renderedMethods;
};

const renderLinks = (apiSpecUrl, apiDocUrl) => {
  if (apiSpecUrl || apiDocUrl) {
    return (
      <div className="d-flex list-regular--item">
        <div className="col-12 pl-0">
          {apiDocUrl && (
            <a href={apiDocUrl} className="mr-5">
              {localization.documentation}
              <i className="fa fa-external-link fdk-fa-right" />
            </a>
          )}
          {apiSpecUrl && (
            <a href={apiSpecUrl} className="mr-5">
              {localization.specification}
              <i className="fa fa-external-link fdk-fa-right" />
            </a>
          )}
        </div>
      </div>
    );
  }
  return null;
};

export const ApiEndpoints = props => {
  const { name, paths, apiSpecUrl, apiDocUrl } = props;
  return (
    <section className="mb-5 list-regular" name={name}>
      <div className="d-flex list-regular--item">
        <div className="col-5 pl-0">
          <h3 className="">{localization.api.endpoints.operation}</h3>
        </div>
        <div className="col-7">
          <h3 className="">{localization.api.endpoints.description}</h3>
        </div>
      </div>
      {renderPaths(paths)}
      {renderLinks(apiSpecUrl, apiDocUrl)}
    </section>
  );
};

ApiEndpoints.defaultProps = {
  name: null,
  paths: null,
  apiSpecUrl: null,
  apiDocUrl: null
};

ApiEndpoints.propTypes = {
  name: PropTypes.string,
  paths: PropTypes.object,
  apiSpecUrl: PropTypes.string,
  apiDocUrl: PropTypes.string
};
