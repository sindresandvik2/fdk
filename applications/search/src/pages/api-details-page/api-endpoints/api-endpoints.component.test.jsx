import React from 'react';
import { shallow } from 'enzyme';
import _ from 'lodash';
import { ApiEndpoints } from './api-endpoints.component';
import apiItem from './__fixtures/apiItem.json';

let defaultProps;

test('should render api-ApiEndpoints correctly with no props', () => {
  const wrapper = shallow(<ApiEndpoints />);
  expect(wrapper).toMatchSnapshot();
});

test('should render api-ApiEndpoints correctly', () => {
  defaultProps = {
    name: 'Test',
    paths: _.get(apiItem, ['openApi', 'paths']),
    apiSpecUrl: _.get(apiItem, 'apiSpecUrl'),
    apiDocUrl: _.get(apiItem, 'apiDocUrl')
  };
  const wrapper = shallow(<ApiEndpoints {...defaultProps} />);
  expect(wrapper).toMatchSnapshot();
});
