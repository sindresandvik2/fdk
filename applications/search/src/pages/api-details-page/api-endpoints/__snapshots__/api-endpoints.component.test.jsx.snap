// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render api-ApiEndpoints correctly 1`] = `
<section
  className="mb-5 list-regular"
  name="Test"
>
  <div
    className="d-flex list-regular--item"
  >
    <div
      className="col-5 pl-0"
    >
      <h3
        className=""
      >
        Operation
      </h3>
    </div>
    <div
      className="col-7"
    >
      <h3
        className=""
      >
        Description
      </h3>
    </div>
  </div>
  <React.Fragment
    key="get-/"
  >
    <div
      className="d-flex list-regular--item"
    >
      <div
        className="col-5 pl-0 fdk-text-strong"
      >
        GET
         
        /
      </div>
      <div
        className="col-7"
      >
        Hent alle tjenester
      </div>
    </div>
  </React.Fragment>
  <React.Fragment
    key="get-/enheter"
  >
    <div
      className="d-flex list-regular--item"
    >
      <div
        className="col-5 pl-0 fdk-text-strong"
      >
        GET
         
        /enheter
      </div>
      <div
        className="col-7"
      >
        Hent alle enheter
      </div>
    </div>
  </React.Fragment>
  <React.Fragment
    key="get-/enheter/{organisasjonsnummer}"
  >
    <div
      className="d-flex list-regular--item"
    >
      <div
        className="col-5 pl-0 fdk-text-strong"
      >
        GET
         
        /enheter/{organisasjonsnummer}
      </div>
      <div
        className="col-7"
      >
        Hent detaljer om enhet
      </div>
    </div>
  </React.Fragment>
  <React.Fragment
    key="post-/underenheter"
  >
    <div
      className="d-flex list-regular--item"
    >
      <div
        className="col-5 pl-0 fdk-text-strong"
      >
        POST
         
        /underenheter
      </div>
      <div
        className="col-7"
      >
        Hent alle underenheter
      </div>
    </div>
  </React.Fragment>
  <React.Fragment
    key="get-/underenheter/{organisasjonsnummer}"
  >
    <div
      className="d-flex list-regular--item"
    >
      <div
        className="col-5 pl-0 fdk-text-strong"
      >
        GET
         
        /underenheter/{organisasjonsnummer}
      </div>
      <div
        className="col-7"
      >
        Hent detaljer om underenhet
      </div>
    </div>
  </React.Fragment>
  <React.Fragment
    key="get-/oppdateringer/enheter"
  >
    <div
      className="d-flex list-regular--item"
    >
      <div
        className="col-5 pl-0 fdk-text-strong"
      >
        GET
         
        /oppdateringer/enheter
      </div>
      <div
        className="col-7"
      >
        Hent oppdateringer på enheter
      </div>
    </div>
  </React.Fragment>
  <React.Fragment
    key="get-/oppdateringer/underenheter"
  >
    <div
      className="d-flex list-regular--item"
    >
      <div
        className="col-5 pl-0 fdk-text-strong"
      >
        GET
         
        /oppdateringer/underenheter
      </div>
      <div
        className="col-7"
      >
        Hent oppdateringer på underenheter
      </div>
    </div>
  </React.Fragment>
  <React.Fragment
    key="get-/organisasjonsformer"
  >
    <div
      className="d-flex list-regular--item"
    >
      <div
        className="col-5 pl-0 fdk-text-strong"
      >
        GET
         
        /organisasjonsformer
      </div>
      <div
        className="col-7"
      >
        Hent alle organisasjonsformer
      </div>
    </div>
  </React.Fragment>
  <React.Fragment
    key="get-/organisasjonsformer/{orgformKode}"
  >
    <div
      className="d-flex list-regular--item"
    >
      <div
        className="col-5 pl-0 fdk-text-strong"
      >
        GET
         
        /organisasjonsformer/{orgformKode}
      </div>
      <div
        className="col-7"
      >
        Hent en enkelt organisasjonsform fra kode
      </div>
    </div>
  </React.Fragment>
  <div
    className="d-flex list-regular--item"
  >
    <div
      className="col-12 pl-0"
    >
      <a
        className="mr-5"
        href="https://data.brreg.no/enhetsregisteret/api/docs/index.html"
      >
        Documentation
        <i
          className="fa fa-external-link fdk-fa-right"
        />
      </a>
      <a
        className="mr-5"
        href="https://raw.githubusercontent.com/brreg/openAPI/master/specs/enhetsregisteret.json"
      >
        Specification
        <i
          className="fa fa-external-link fdk-fa-right"
        />
      </a>
    </div>
  </div>
</section>
`;

exports[`should render api-ApiEndpoints correctly with no props 1`] = `
<section
  className="mb-5 list-regular"
  name={null}
>
  <div
    className="d-flex list-regular--item"
  >
    <div
      className="col-5 pl-0"
    >
      <h3
        className=""
      >
        Operation
      </h3>
    </div>
    <div
      className="col-7"
    >
      <h3
        className=""
      >
        Description
      </h3>
    </div>
  </div>
</section>
`;
